File: leetcode/CPP/125-Valid-Palindrome.cpp

/*125. Valid Palindrome
Given a string, determine if it is a palindrome, considering only alphanumeric 
characters and ignoring cases.
*/
bool isPalindrome(string s)
{
    int n = s.length();
    int l = 0, r = n - 1;
    while (l < r)
    {
        if (!isalnum(s[l]))
            l++;
        else if (!isalnum(s[r]))
            r--;
        else
        {
            if (tolower(s[l]) != tolower(s[r]))
                return false;
            l++;
            r--;
        }
    }
    return true;
}

File: leetcode/python/125-Valid-Palindrome.py
'''
125. Valid Palindrome
Given a string, determine if it is a palindrome, considering only alphanumeric 
characters and ignoring cases.
'''
# Two Pointer
def isPalindrome(s):
    i = next_of_alnum(s, 0)
    j = prev_of_alnum(s, len(s) - 1)
    while (i < j):
        if (s[i].lower() != s[j].lower()):
            return False
        else:
            i, j = next_of_alnum(s, i + 1), prev_of_alnum(s, j - 1)
    return True
    

def next_of_alnum(s, pos):
    while (not s[pos].isalnum()):
        pos += 1
        if pos == len(s):
            return len(s)
    return pos

def prev_of_alnum(s, pos):
    while (not s[pos].isalnum()):
        pos -= 1
        if pos == -1:
            return -1
    return pos
